network/p2p/middleware.go

```diff
@@ -72,17 +82,24 @@ type Middleware struct {
 
 // NewMiddleware creates a new middleware instance with the given config and using the
 // given codec to encode/decode messages to our peers.
-func NewMiddleware(log zerolog.Logger, codec network.Codec, address string, flowID flow.Identifier,
-	key crypto.PrivateKey, metrics module.NetworkMetrics, maxUnicastMsgSize int, maxPubSubMsgSize int,
-	rootBlockID string, validators ...network.MessageValidator) (*Middleware, error) {
+func NewMiddleware(log zerolog.Logger,
+	ctx context.Context,
+	cancel context.CancelFunc,
+	libP2PHostGenFunc LibP2PFactoryFunc,
+	codec network.Codec,
+	address string,
+	flowID flow.Identifier,
+	key crypto.PrivateKey,
+	metrics module.NetworkMetrics,
+	maxUnicastMsgSize int,
+	maxPubSubMsgSize int,
+	rootBlockID string,
+	validators ...network.MessageValidator) (*Middleware, error) {
 	ip, port, err := net.SplitHostPort(address)
 	if err != nil {
 		return nil, fmt.Errorf("failed to create middleware: %w", err)
 	}
 
-	p2p := &Node{}
-	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
To me calling `WithCancel` within the constructor was more flexible. If the caller also needs a `cancel` function, it's fine to create a child context there as well, but we don't need to force the caller to create and pass in a `cancel`. It's also a bit unusual to allow the holder of the child context to cancel higher-level contexts -- the pattern is the other way around.

Reply:
fixed in https://github.com/onflow/flow-go/pull/195/commits/2d17425d0e6f723affd43c1daae7ea08b91c8d6a.